lib/spack/llnl/util/lang.py

```diff
@@ -1018,3 +1034,49 @@ def __init__(self, callback):
 
     def __get__(self, instance, owner):
         return self.callback(owner)
+
+
+class DeprecatedProperty:
+    """Data descriptor to error or warn when a deprecated property is accessed.
+
+    Derived classes must define a factory method to return an adaptor for the deprecated
+    property, if the descriptor is not set to error.
+    """
+
+    __slots__ = ["name"]
+
+    #: 0 - Nothing
+    #: 1 - Warning
+    #: 2 - Error
+    error_lvl = 0
+
+    #: Whether to add tracebacks to warnings and errors
+    traceback = False
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+
+    def __get__(self, instance, owner):
+        if instance is None:
+            return self
+
+        if self.error_lvl == 1:
+            msg = f"accessing the '{self.name}' property of '{instance}', which is deprecated"
+            warnings.warn(msg)
+            if self.traceback:
+                traceback.print_stack()
```

Comment:
I also feel like this doesn't really belong here. I think the warning hook can deal with it (?)

Reply:
Fair, I guess we can tweak warnings.showwarning if we need that